yoga/node/LayoutableChildren.h

```diff
@@ -0,0 +1,142 @@
+        : node_(node),
+          childIndex_(childIndex),
+          backtrack_(std::move(backtrack)) {}
+
+    T* operator*() const {
+      return node_->getChild(childIndex_);
+    }
+
+    Iterator& operator++() {
+      next();
+      currentNodeIndex_++;
+      return *this;
+    }
+
+    Iterator operator++(int) {
+      Iterator tmp = *this;
+      ++(*this);
+      return tmp;
+    }
+
+    size_t index() const {
+      return currentNodeIndex_;
+    }
+
+    friend bool operator==(const Iterator& a, const Iterator& b) {
+      return a.node_ == b.node_ && a.childIndex_ == b.childIndex_;
+    };
+
+    friend bool operator!=(const Iterator& a, const Iterator& b) {
+      return a.node_ != b.node_ || a.childIndex_ != b.childIndex_;
+    };
+
+   private:
+    void next() {
+      if (childIndex_ + 1 >= node_->getChildCount()) {
+        // if the current node has no more children, try to backtrack and
+        // visit its successor
+        if (backtrack_.empty()) {
+          // if there are no nodes to backtrack to, the last node has been
+          // visited
+          node_ = nullptr;
+          childIndex_ = SIZE_MAX;
+        } else {
+          // pop and restore the latest backtrack entry
+          const auto back = backtrack_.back();
+          backtrack_.pop_back();
+          node_ = back.first;
+          childIndex_ = back.second;
+
+          // go to the next node
+          next();
+        }
+      } else {
+        // current node has more children to visit, go to next
+        ++childIndex_;
+        // skip all display: contents nodes, possibly going deeper into the
+        // tree
+        skipContentsNodes();
+      }
+    }
+
+    void skipContentsNodes() {
+      // get the node that would be returned from the iterator
+      auto currentNode = node_->getChild(childIndex_);
+      while (currentNode->style().display() == Display::Contents &&
+             currentNode->getChildCount() > 0) {
+        // if it has display: contents set, it shouldn't be returned but its
+        // children should in its place push the current node and child index
+        // so that the current state can be restored when backtracking
+        backtrack_.push_back({node_, childIndex_});
+        // traverse the child
+        node_ = currentNode;
+        childIndex_ = 0;
+
+        // repeat until a node without display: contents is found in the
+        // subtree or a leaf is reached
+        currentNode = currentNode->getChild(childIndex_);
+      }
+
+      // if no node without display: contents was found, try to backtrack
+      if (currentNode->style().display() == Display::Contents) {
+        next();
+      }
+    }
+
+    const T* node_;
+    size_t childIndex_;
+    size_t currentNodeIndex_{0};
+    Backtrack backtrack_;
+
+    friend LayoutableChildren;
+  };
+
+  LayoutableChildren(const T* node) : node_(node) {
+    static_assert(std::input_iterator<LayoutableChildren<T>::Iterator>);
```

Comment:
Apparently there was a C++ 20 standard defect that led older stdlib to require these to be default constructible 😮. https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2325r3.html

Going to locally patch in a default ctor that just initializes to nullptr...

```
buck-out/v2/gen/fbsource/aaebd9ffd9ca7aaa/xplat/yoga/__yogaFbcode__/buck-private-headers/yoga/node/Node.h:150:12: note: in instantiation of member function 'facebook::yoga::LayoutableChildren<facebook::yoga::Node>::LayoutableChildren' requested here
    return LayoutableChildren(this);
           ^
buck-out/v2/gen/fbsource/aaebd9ffd9ca7aaa/xplat/yoga/__yogaFbcode__/buck-private-headers/yoga/node/LayoutableChildren.h:118:19: note: because 'LayoutableChildren<Node>::Iterator' does not satisfy 'input_iterator'
    static_assert(std::input_iterator<LayoutableChildren<T>::Iterator>);
                  ^
fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/iterator_concepts.h:634:30: note: because 'facebook::yoga::LayoutableChildren<facebook::yoga::Node>::Iterator' does not satisfy 'input_or_output_iterator'
    concept input_iterator = input_or_output_iterator<_Iter>
                             ^
fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/iterator_concepts.h:614:5: note: because 'facebook::yoga::LayoutableChildren<facebook::yoga::Node>::Iterator' does not satisfy 'weakly_incrementable'
        && weakly_incrementable<_Iter>;
           ^
fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/iterator_concepts.h:597:36: note: because 'facebook::yoga::LayoutableChildren<facebook::yoga::Node>::Iterator' does not satisfy 'default_initializable'
    concept weakly_incrementable = default_initializable<_Iter>
                                   ^
fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/concepts:143:37: note: because 'facebook::yoga::LayoutableChildren<facebook::yoga::Node>::Iterator' does not satisfy 'constructible_from'
    concept default_initializable = constructible_from<_Tp>
                                    ^
fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/concepts:139:30: note: because 'is_constructible_v<facebook::yoga::LayoutableChildren<facebook::yoga::Node>::Iterator>' evaluated to false
      = destructible<_Tp> && is_constructible_v<_Tp, _Args...>;
```